src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -511,14 +521,153 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol
                     set.Free();
                     return false;
             }
-            
+
             // very rare case. 
             // some implemented interfaces are related
             // will have to instantiate interfaces and check
             hasRelatedInterfaces:
             return type.InterfacesNoUseSiteDiagnostics(basesBeingResolved).HasDuplicates(TypeSymbol.EqualsIgnoringDynamicAndTupleNamesComparer);
         }
 
+        /// <summary>
+        /// Deep checking tuple names means checking the constraints on every constituent type and all their base types.
+        /// deepCheckTupleNames should be set to false when the caller is already visiting constituent types.
+        /// </summary>
+        private static bool CheckTupleNamesConstraints(TypeSymbol type, ArrayBuilder<TypeParameterDiagnosticInfo> builder, TupleNamesCheckKind deepCheckTupleNames)
+        {
+            // Note this only checks for tuple name errors introduced by referencing this type.
+            //
+            // So if the type does not include any tuples, its reference cannot introduce an error.
+            // But its declaration could. For instance:
+            //     `class Error : Base, I<(int notA, int notB)>`
+            //
+            // Similarly, if the base type doesn't use type arguments from the type, then the reference of this type cannot introduce an error.
+            // But the base type may already have one (reported elsewhere). For instance:
+            //     `class C : BaseCanBeSkipped<int>, ITest`
+            //     `class D : BaseNeedsChecking<T>, ITest`
+            if (!type.ContainsTuple())
+            {
+                return true;
+            }
+            switch (deepCheckTupleNames)
+            {
+                case TupleNamesCheckKind.Shallow:
+                    {
+                        if (!CheckTupleNamesConstraintsCore(type, builder))
+                        {
+                            return false;
+                        }
+
+                        // Walk up the base types and apply the same check on each inheritence level.
+                        // If the base type was not "variable", then we can cut short.
+                        TypeSymbol currentType = type;
+                        NamedTypeSymbol baseType = currentType.BaseTypeNoUseSiteDiagnostics;
+                        bool result = true;
+                        while ((object)baseType != null && currentType.OriginalDefinition.BaseTypeNoUseSiteDiagnostics.ContainsTypeParameter())
+                        {
+                            result &= CheckTupleNamesConstraints(baseType, builder, TupleNamesCheckKind.Shallow);
```

Comment:
That was indeed wrong. Added test in iteration 7. Thanks

---
In reply to: [93093404](https://github.com/dotnet/roslyn/pull/15809#discussion_r93093404) [](ancestors = 93093404)